### PR TITLE
Custom bar colors

### PR DESCRIPTION
**Pull request by GRFreire**
Set the bar colors independently from the bg/fg colors.

`.Xresources` (example):
```
Sxiv.bar: black
Sxiv.text: white
```

**Notes:**
Bar color and text color defaults to foreground and background colors respectively, so it wont break backwards compatibility.

I'm not sure about the name we should use in the Xresources file, it could be `bar/textColor` but since background/foreground does not have this suffix, I preferred not to.

### CODE REVIEW

**Comment by eylles** [comment]
i may be biased as this is how i did it in rxiv but i'd suggest naming the resources like this:
```c
.bar.background
.bar.foreground
```

i have 2 rationales for why i think this is the more correct way:

1. the usual structure for xresources is `Program.resource: value` or `Program.class.resource: value`, this way the bar resources would adhere to the second standard as:
```c
Nsxiv.bar.background: #140C16
Nsxiv.bar.foreground: #CEB9D0
```

2. if you generate your xresources colors or like to set everything to have the same background and foreground colors then you do something like this:
```c
*.foreground:       #c1c0c1
*.background:       #080507
```
these globs would then match for Nsxiv.bar.background and Nsxiv.bar.foreground in effect giving the bar the same colors as the nsxiv window.

**Comment by GRFreire** [comment]
> ```c
> *.foreground:       #c1c0c1
> *.background:       #080507
> ```
> 
> these globs would then match for Nsxiv.bar.background and Nsxiv.bar.foreground in effect giving the bar the same colors as the nsxiv window.

but to be able to do this the window background/foreground should also follow the second standard (`Program.class.resource: value`), like this:
```
Nsxiv.window.background: #140C16
Nsxiv.window.foreground: #CEB9D0
```
and thus breaking backwards compatibility. Am I right or am I missing something here?

Apart from that, what the other maintainers think of this? The current way I've already coded or the @eylles suggestion? In my opinion the current way is better and simpler, even tho I found your idea interesting (didn't even know about this before actually )

**Comment by XPhyro** [comment]
> but to be able to do this the window background/foreground should also follow the second standard (`Program.class.resource: value`), like this:
> 
> ```
> Nsxiv.window.background: #140C16
> Nsxiv.window.foreground: #CEB9D0
> ```
> 
> and thus breaking backwards compatibility. Am I right or am I missing something here?

I think it should be *bar*/*text* because of this. If we comply with the standard, we either become inconsistent or break backwards compatibility.

**Comment by GRFreire** [comment]
> Since this is the beginning of nsxiv, we could choose to also update the window properties. People would have to update their Xresources regardless as they'll have to write `Nsxiv` instead of `Sxiv`.

That makes sense. But afaik in the main branch we still set options with `Sixv.*`. Should we change this (to `Nsxiv`) in this pr or it is out of scope? Maybe #9 takes care of this, but it will take some time until it's merged.

Before we proceed, anyone against changing the Xresources variables as described by @eylles ? All fine by me and apparently @XPhyro and @eylles 
```
Nsxiv.window.background: #140C16
Nsxiv.window.foreground: #CEB9D0
Nsxiv.bar.background: #0E0E0E
Nsxiv.bar.foreground: #FCFCFC
```

**Comment by XPhyro** [comment]
> That makes sense. But afaik in the main branch we still set options with `Sixv.*`. Should we change this (to `Nsxiv`) in this pr or it is out of scope? Maybe #9 takes care of this, but it will take some time until it's merged.

Indeed, #9 takes care of it. @eylles recommended in #1 that the renaming should be done after merging the initial patches by their experience from rxiv.

**Comment by GRFreire** [comment]
Yes, that is why I am asking what should we do. Wait for #9 be merged (that will take some time) or change it here so we can already close this pr.

**Comment by eylles** [comment]
We should be able to merge this first without issue, if you take a look the Program part of the Program.class.resource construct ain't defined here, as that is defined iirc from the name of the program itself, so if we merge this it will search for `Sxiv.window.background`, `Sxiv.window.foreground`, `Sxiv.bar.background`, `Sxiv.bar.foreground`.
And when #9 gets merged each instance of Sxiv will get changed to Nsxiv, so all we have to do here is decide on the name of the resources without minding the `Program` part on this very pr.

**Comment by XPhyro** [comment]
I think this PR can be merged regardless. The program name is defined via `RES_CLASS`, so there is no conflict.

This PR should just change the parts after the program name as everything, including the generated executable, assumes the name sxiv. I don't think it makes sense to merge this specific case before the total rename.

**Comment by GRFreire** [comment]
Ok. Tomorrow I am implementing these changes https://github.com/nsxiv/nsxiv/pull/10#issuecomment-917309996 then

File: sxiv.1
Comment (N-R-K) [inline sxiv.1:391]: I would mention that they default to window foreground/background in here.

File: sxiv.h
Comment (N-R-K) [inline sxiv.h:413]: Lets keep `bg` and `fg` as is and name the bar colors `bar_fg` and `bar_bg`.

File: window.c
Comment (N-R-K) [inline window.c:125]: As mentioned by @eylles , `.bar.foreground` and `.bar.background` would be better.

**Review by N-R-K — CHANGES_REQUESTED** [review]

**Comment by GRFreire** [comment]
> I would mention that they default to window foreground/background in here.

About that, the way is currently working is that bar.background defaults to window.foreground, cause sxiv by default used to have the colors of the bar as the inverted of the window. Is it ok to keep like that or is it better to bar.background default to window.background?

File: sxiv.h
Comment (GRFreire) [reply sxiv.h:413]: Idk, what the other people think about that. For me is makes the code more readable this way, and its not that big of a change.

**Comment by N-R-K** [comment]
I would value not needlessly breaking backwards compatibility.

**Comment by GRFreire** [comment]
What I am saying is that, by changing the current xresources to:
```
Sxiv.window.background: #140C16
Sxiv.window.foreground: #CEB9D0
Sxiv.bar.background: #0E0E0E
Sxiv.bar.foreground: #FCFCFC
```

like I [previously said](https://github.com/nsxiv/nsxiv/pull/10#issuecomment-917320648) I was going to change, it will break backwards compatibility, meaning that people wont be able to change the window color - as they used to - like this:
```
Sxiv.background: #140C16
Sxiv.foreground: #CEB9D0
```

Sorry if I was not making myself clear, but I think its well explained now.

File: sxiv.h
Comment (N-R-K) [reply sxiv.h:413]: Since the scope of this pr is to add custom bar colors, we should leave anything else untouched.
If other's also think new names are more readable, then it should be it's own commit imo.

**Comment by N-R-K** [comment]
Oh, well my comment about not breaking backwards compatibility still remains.
Let's keep `Sxiv.background` and `Sxiv.foreground` as is. And change bar colors to `Sxiv.bar.background` and `Sxiv.bar.foreground`.

**Comment by eylles** [comment]
i was checking and xterm does have resources defined in the form `Program.resource` as well as `Program.class.resource` so keeping `Sxiv.background` and `Sxiv.foreground` while adding `Sxiv.bar.background` and `Sxiv.bar.foreground` won't truly be inconsistent to how other programs use and define resources, altho someone may argue xterm is not the best program to use as reference for some things...

**Comment by N-R-K** [comment]
> i was checking and xterm does have resources defined in the form `Program.resource` as well as `Program.class.resource` so keeping `Sxiv.background` and `Sxiv.foreground` while adding `Sxiv.bar.background` and `Sxiv.bar.foreground` won't truly be inconsistent to how other programs use and define resources, altho someone may argue xterm is not the best program to use as reference for some things...

Do we want to keep `Sxiv` prefix? If yes then we shouldn't break backwards compatibility, if not then lets just do `Nsxiv.window`. I personally don't oppose changing to `Nsxiv`.

Also @GRFreire if we're going for `Nsxiv` I think having `.bar.background` default to `.window.background` would be more sensible.

**Comment by eylles** [comment]
> 
> Do we want to keep `Sxiv` prefix? If yes then we shouldn't break backwards compatibility, if not then lets just do `Nsxiv.window`. I personally don't oppose changing to `Nsxiv`.
> 
> Also @GRFreire if we're going for `Nsxiv` I think having `.bar.background` default to `.window.background` would be more sensible.

the change of the program name in the resource assignement is handled by the rename to nsxiv pr, so it falls outside the scope of this pr, if that's what you mean tho.

**Comment by GRFreire** [comment]
> Also @GRFreire if we're going for Nsxiv I think having .bar.background default to .window.background would be more sensible.

Makes sense, I'll change that.

> Do we want to keep Sxiv prefix? If yes then we shouldn't break backwards compatibility, if not then lets just do Nsxiv.window. I personally don't oppose changing to Nsxiv.

As stated [here](https://github.com/nsxiv/nsxiv/pull/10#issuecomment-917319757), #9 should take care of that renaming. We could change `RES_CLASS` in this pr, it would work and probably wont have any merge conflicts with pr 9, but, like I said, don't know if it is out of scope.

But I agree with you, if we are keeping `Sxiv.` we shouldn't break backwards compatibility. What do you think? Rename `RES_CLASS`, keep `Sxiv.background/foreground` the same, do as @eylles suggested (like xterm). I honestly dont know.

**Comment by N-R-K** [comment]
I think it's fine letting the rename PR https://github.com/nsxiv/nsxiv/pull/9 take care of changing `RES_CLASS`. Seems more appropriate that way.

The main question here should be do we keep `Sxiv` prefix or move to `Nsxiv` ? I'm leaning towards `Nsxiv` but am fine if others think otherwise.

**Comment by GRFreire** [comment]
`RES_CLASS` is what defines that prefix

**Comment by N-R-K** [comment]
> `RES_CLASS` is what defines that prefix

Yes, I'm aware. What I meant was that weather we decide `Sxiv` or `Nsxiv` the rename PR can take care of that.

**Comment by GRFreire** [comment]
I think the rename PR is OK, it should rename `Sxiv` to `Nsxiv`. What I was concerned about is that #9 is probably going to take some time to be merged. So, the time between this PR and the rename PR gets merged the xresources will keep the `Sxiv.` prefix but is still going to break BC (because of `Sxiv.window.background`).

Is that OK? If it is, then this PR is ready to be accepted (i think)

**Comment by N-R-K** [comment]
We can wait a bit then. Also I think we should merge all the smaller patches and then roll the rename PR out. Some of the bigger patches seems like they'll need heavy refactoring anyways.

**Comment by eylles** [comment]
for me it is ok as the 2 breaking changes are documented, so anyone that migrates to this will be aware, perhaps it would be a good idea to notify the breaking changes on the changelog?

File: sxiv.h
Comment (N-R-K) [reply sxiv.h:413]: @GRFreire are you fine with `win_bg`, `bar_bg` as var names? This mirrors the Xresources entries so I think it'd be more appropriate.

**Comment by GRFreire** [comment]
Wait a sec, just realized that this PR is target at `nsxiv:dev`, sooo there is no problem in merging it right away. Just make sure the rename pr is accepted before the new version is released (and `dev` is merged into `master`)

**Comment by GRFreire** [comment]
> perhaps it would be a good idea to notify the breaking changes on the changelog?

sure

File: sxiv.h
Comment (eylles) [reply sxiv.h:413]: all left is to resolve this and the pr can be merged.

Comment (GRFreire) [reply sxiv.h:413]: Yeah, changing to `win_bg/fg` and `bar_bg/fb` then

Comment (N-R-K) [reply sxiv.h:413]: I think we're also changing the default bar color values.

Comment (GRFreire) [reply sxiv.h]: Yes, changed that on my last commit

**Review by XPhyro — APPROVED** [review]

**Review by N-R-K — APPROVED** [review]
LGTM

**Comment by GRFreire** [comment]
Quick question, what is your preference about merge? Squash, rebase or merge commit?

**Comment by eylles** [comment]
I think that for this kind of stuff that needs multiple commits to really nail down the patch a squash is cleaner

**Comment by N-R-K** [comment]
My view is :
* squash - for multiple commits. A single PR should serve a single purpose, so squashing makes things cleaner.
* rebase - for single commits where the mainline branch has moved further.

**Pull request merged**
